### ISSUE TITLE
withStyle helper method

### ISSUE THREAD

**Issue by marzelin** [paypal/glamorous#145]
Let's say you want a red button. What would be the most straight forward way you can think of?
What about:

```javascript
const Tomato = Button.withStyle({backgroundColor: 'tomato'})
```
It feels so natural, reads like English and does exactly what it says.

OK, you've got a boring tomato, let's make it more interesting:
```javascript
const HoverableTomato = Tomato.withStyle({'&:hover': {color: 'goldenrod'}})
```
Exactly the same method as before. Finally one simple and obvious way of doing things.

You think it's too good to be true? It's here! Check it out on [codesandbox.io](https://codesandbox.io/s/ERLK5X2nN)

Styling in this way feels so good. :)

**Comment by ajwhite** [comment]
You can use the `glamorous` "factory factory" function 

```js
const Tomato = glamorous(Button)({backgroundColor: 'tomato'})
```
```js
const HoverableTomato = glamorous(Tomato)({'&:hover': {color: 'goldenrod'}})
```

And continue the HOC-inception 😃 

I do like the idea of `withStyle` though as a convenient way to accomplish the same thing

**Comment by marzelin** [comment]
It's all about user convenience.
Why somebody would want to use this library? To improve their styling workflow.
What is a user is welcomed with? 17-page-long documentation.
I can imagine somebody looking at the docs and running away to their crappy but familiar workflow. "I just wanted add some styles, not get phd in computer science. Thankyouverymuch"

Let's get out of user's way as much as possible. Let them do the thing their want in as much straight-forward way there is.
Do you want an html element with some styles? Here you go: `HtmlElement.withStyle(someCssRules)`
 There's no mental translation needed. What you want is what you write. And inheriting from custom components is done in the same way - one thing less to learn and remember which converts to more time spent on actual work.

**Comment by marzelin** [comment]
## Reasoning behind this change
### Interoperability
When you use current api, you have to keep in mind whether you're working with a `componentFactory` or `Component`. If you try to use `componentFactory` in jsx or try to style `Component` your application blows up.

With new api, when you import a built-in component
```javascript
import { Div } from 'glamorous'
```

You can use it directly in jsx
```javascript
<Div>Hello</Div>
```
Style it in jsx
```javascript
<Div css={{color: 'white'}}>Hello</Div>
```

Style it in js to benefit from reusability and inheritance
```javascript
const WhiteDiv = Div.withStyle({color: 'white'})
```
The user don't have to care about `componentFactory`/`Component` differences anymore. The library handles it internally for them.

### Readability (aka clear intention)
When you ask somebody who never heard of `glamorous` what this code does:
```javascript
const Primary = glamorous.div({color: 'tomato'})
```
the respondent would probably be confused.

What about this:
```javascript
const Primary = Div.withStyle({color: 'tomato'})
```
Primary is a div with style that sets its color to tomato?

### invisibility === success
When you look at successful libraries, one thing that is quite common is that those libraries try to be as much out of sight as possible - users don't want to type a library name a hundred times.
When you use lodash, it hides behind `_`, jQuery uses `$`. mocha gives users just the things they care about: `describe` and `it` methods. Glamorous is an awesome name, but it should be used only once per file - in `import` statement.

### Improved maintability
When we only expose the `Component` api with `withStyle` helper we would be able to simplify codebase which means less overhead with maintaining the library. As a user I don't care about ten ways of doing the same thing - I'd love to have one consistent, convenient and predictable way.

**Comment by kentcdodds** [comment]
I have thoughts about this and will share them when I have more time. Thanks 😀

**Comment by ajwhite** [comment]
@marzelin I totally get the motivation, thanks for writing up an explanation of your thoughts!

I'm curious to hear @kentcdodds's thinking as well. There's similar APIs being discussed in https://github.com/paypal/glamorous/issues/133 and we'll want to make sure these coincide together nicely.

**Comment by kentcdodds** [comment]
Hey @marzelin,
Thank you very much for sharing your thoughts and arguments. I can appreciate your perspective and understand what you mean about the API being a bit more straightforward. The trade-off of doing something like this is that now there's two ways to do the same thing:

```javascript
const MyComponent = (...props) => <div {...props} />
const MyStyledComponent = glamorous(MyComponent)({})

// these two are the exact same
const MyStyledComponentWithStyles = glamorous(MyStyledComponentWithStyles)({})
const MyStyledComponentWithStyles2 = MyStyledComponent.withStyles({})
```

The problem is that if folks learn the `.withStyles` API, then when they want to style a component that wasn't created by glamorous they'll not know how to accomplish that.

However, if folks learn the `glamorous` API, they know how to style a glamorous component or one that wasn't created by glamorous. It's pretty straightforward.

Does that make sense? So I'm leaning toward not implementing this.

**Comment by kentcdodds** [comment]
I'm starting to think this might be good actually... Still thinking it over. I'd be willing to look at a pull request I think.

**Comment by marzelin** [comment]
You can wrap any component in `glamorous()` but it only work if the component properly uses `className` that is passed to it. If it doesn't you get no error but your code doesn't work as expected. Good luck with debugging.

Let's say you have this line in your project
```javascript
const StyledPosts = glamorous(Posts)
```
And your coworker changes `Posts` component in a way that prevents styles from being propagated in a desired way. They gets no error whatsoever. Unless you have robust tests you're in trouble.

Also this is against the convention:
```javascript
<Container_Component>
    <Presientational_Component/>
</Container_Component>
```
You want `containers` to wrap around `presentationals`, not the other way around.

Friends don't let friends style non-glamorous components because it's a possible source of bugs, so please don't do it :)

To sum up: I find errors on trying to style non-glamorous components really helpful.
It makes me think of a better way of doing what I want to implement and if I don't find one, to double-check if the component uses `className` correctly and provide proper info for anyone who would try to change it.

**Comment by kentcdodds** [comment]
What are you suggesting @marzelin? That we somehow log an error if the provided `className` is not used correctly? I'm not sure that's possible 🤔

**Comment by marzelin** [comment]
@kentcdodds I suggest stop using things that are not 100% reliable if there's a better alternative.

**Comment by kentcdodds** [comment]
As discussed here: https://github.com/paypal/glamorous/issues/133#issuecomment-304524601 maybe it would be better for us to make this be out of band.

> I suggest stop using things that are not 100% reliable if there's a better alternative.

What do you mean? What are you suggesting we change? Also, I should say that I _really_ want to avoid being dogmatic about anything in the docs. We can suggest things and mention trade-offs/foot guns/etc. But I really don't want to tell people that they should never do one thing or another.

**Comment by marzelin** [comment]
I think `glamorous()` function is unsafe to use because it fails silently when a component doesn't handle `className` properly. I'd deprecate it to save people from rare but nasty bugs. But if there will be a warning that `glamorous()` function is super convenient but can bite you in the neck if you're not cautious I'm okay with that.

I'm ok with moving `withStyle` to a separate helper but only if it will be able to check if a passed component is a GlamorousComponent an throws and error if not.

But for me
```javascirpt
Div.withStyle({color: 'AntiqueWhite'})
```
looks much cooler than
```
withStyle(Div, 'AntiqueWhite')
```

**Comment by marzelin** [comment]
Another big win of `withStyle` as a method is that if I import a component and want to change its style a bit, I wouldn't have to import `glamorous` to that file just for that - `.withStyle` will work out of the box.
Importing `react` in every file where jsx is present sooo annoying :angry:

**Comment by ErikCupal** [comment]
I have rather conservative opinion about this. I think that when there are two things in a library that are doing pretty much the same thing, it causes more confusion, especially to people, who don't know the library good enough 🙁.

**Comment by marzelin** [comment]
That's a valid point.
The question is which api do you prefer:
Current, familiar but failing silently in some cases:
```javascript
onst Tomato = glamorous(Button)({backgroundColor: 'tomato'})
```
Or this
```
const Tomato = Button.withStyle({backgroundColor: 'tomato'})
```
that behaves consistently: either work or let you know that something is wrong.

Changing primary function would be a brave decision but could give us a serious advantage over `styled-components` that share this weakness: `glamorous` will be reliable, `styled-components` are not.

**Comment by kentcdodds** [comment]
Still not at the keyboard, but I just have to say that there's no way we'll be deprecating the existing functionality of the glamorous function. I'll justify that decision when I have more time later.

**Comment by kentcdodds** [comment]
> Current, familiar but failing silently in some cases

Yes, it will fail silently in some cases. To be specific, if you don't apply the `className` to the underlying rendered element. However, there are _many_ use cases in which we cannot force folks to apply the `className` _and_ it's impossible to ensure that they do. And we wouldn't want to warn if they don't because it could be that the element is not rendered at all for a good reason. In addition, the `className` could be applied to any element (by design), not just the root element of the component. So it's not reasonable/possible/correct for us to force folks to apply the `className`.

Another reason that we'll not be deprecating the capabilities of the `glamorous` function is because one use case that we really want to support is the ability to style a third party component. For example, I use the [`svg-inline-react`](https://github.com/sairion/svg-inline-react) component and use glamorous to style that. This is a very common use case and is only possible because of the `glamorous` API.

> Another big win of withStyle as a method is that if I import a component and want to change its style a bit, I wouldn't have to import glamorous to the file just for that - .withStyle will work out of the box.
> Importing react into every file where jsx is present sooo annoying 😠

I personally prefer the explicitness and this is not actually a drawback of the `glamorous` function to me, but a feature.

Another drawback of the `withStyle` function being dangled on a glamorous component is that you're never certain whether what you're importing is a glamorous component or not (and whether it supports the special APIs of glamorous). I've noticed this as I've been using glamorous. It's already a bit of a frustration for me to know whether I can use the `css` prop on a component. I can't know without looking at the implementation. It'd prefer to just use `glamorous` on the component and not worry about it at all.

I think that the `className` forwarding requirement is something we'll have to live with and document well (note #112).

I'm going to go ahead and close this, but as always with OSS, I can be convinced otherwise and we could reopen in the future.

**Comment by marzelin** [comment]
Yeah, there's not enough enthusiasm that would justify introducing this change anyway (and nobody shared my excitement :cry:). I'm considering forking this project to give people like myself an alternative api. Would you mind If I do this?

**Comment by kentcdodds** [comment]
I wonder if we could make a plugin system that would allow you to create a plugin for it and not have to fork the project (and community). This would also allow us to try things out before including them in the library officially. Could you open an issue about that and we can discuss what that would look like?
